src/storage/mod.rs

```diff
@@ -78,10 +69,37 @@ pub enum StorageError {
     Unknown(String),
 }
 
+/// A trait that represents whether some struct can be read from a given store
+///
+/// Note that these msut be this way because we can't write non-overlapping blanked impls yet:
+/// https://github.com/rust-lang/rust/issues/20400
+/// https://stackoverflow.com/questions/40392524/conflicting-trait-implementations-even-though-associated-types-differ
+#[async_trait]
+pub trait StoreRead<State,Id> {
+    /// Read a given piece of state from the store
+    async fn read(&self, id: Id) -> Result<State, StorageError>;
+}
+
+/// A trait that represents whether some struct can be written to a given store
+#[async_trait]
+pub trait StoreWrite<State,Id> {
+    /// Write to a given store
+    async fn write(&mut self, obj: State, id: Option<Id>) -> Result<Id, StorageError>;
```

Comment:
Is there a reason why we wouldn't ever need an ID? Feels like a cleaner API for this not to be an option

Reply:
This was in the case that you *don't know* the ID ahead of time -- like if you're storing something and you're going to get your ID from the storage layer.

Could be premature, since just about every storage layer will probably be happy with a uuidv1/4/7/8